Add Best Practice: Do not hardcode task names

liutikas opened the pull request:
Docs tested with: `./gradlew serveDocs -PquickDocs`

Tests ran with `./gradlew :docs:docsTest --tests "*snippet-best-practices-avoid-hardcoding-*"`

<!--- The issue this PR addresses -->
<!-- Fixes #? -->

### Context
<!--- Why do you believe many users will benefit from this change? -->
<!--- Link to relevant issues or forum discussions here -->

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

liutikas committed:
Add a best practice to not hardcode task names

Docs tested with: ./gradlew serveDocs -PquickDocs

Tests ran with ./gradlew :docs:docsTest --tests "*snippet-best-practices-avoid-hardcoding-*"

Signed-off-by: Aurimas Liutikas <aurimas@google.com>

tresat reviewed (changes requested):
So I've talked this item over with @big-guy, and there are some clarifications we want to make before getting into more detailed wordsmithing.

## The Explanation

> Task names are consider implementation details of a plugin

This is sometimes true and sometimes not.  Many Gradle core plugins expose the task name as a static constant on the task class or on the plugin class - these would be considered public API and would require a full deprecation cycle if we were ever to change them (this is very unlikely).  This may be more true in the Android world, but it's not a universal principle by any means.

Relatedly, there is no iron law that says that task _types_ have to remain the same from version to version of a plugin.  These are also likely to be seen by plugin authors as implementation details in many cases, and might be refactored on a whim.  Changes here at least have the advantage of failing fast when a buildscript uses the old name and fails to compile - but this is still annoying.

> Making assumptions about task names makes a build fragile to Gradle and plugin updates.

This only applies if names are _not_ part of the plugin's public API.

There seem to be 2 different personas we're talking to here:

### As a Plugin Author:

Don't force your users to rely on task names _or_ task types.  Provide a DSL extension block for configuring behavior that is then wired into the appropriate tasks your plugin creates.  This allows consumers of your plugin to work at a higher level of abstraction, using a more declarative approach.  Don't expose task names as part of your public API as static constants on task or plugin types.

### As a Build Author:

Consider the best way to access a task.  Prefer Plugin DSL > Configuring Tasks by Type > Task Names.  Be aware that only the first should really be considered future-proof (and when we start writing more plugin author best practices, "Don't Change Your Extensions (including their names) without a deprecation cycle" should be one of them).  Sometimes plugin DSL won't be available and you'll have to use types.  Sometimes there will be multiple instances of the same type and you'll only want to configure some of them so you might be forced to use names.

In an ideal world, a recommendation when running builds might be to only execute _lifecycle tasks_ which lack any configuration except task dependencies.  The name and type details of the configurable tasks that actually do work would then be hidden.  This would however require Plugin Authors to reliably create such tasks, and wire their own tasks into existing lifecycle tasks as dependencies.

That aside, we need to make this item a little more nuanced.  We don't want a universal recommendation against calling `named()`.

## The Example

A problem with the example is that the Avoid and Do work completely differently for the `test` configuration example, as noted in the bullet point.  In the Avoid, we are intending to configure only a single task.  In the Do, we are configuring every task of that type _including any that may be added by other plugins applied later_.  These are very different behaviors and assuming this will only configure the _current_ or _desired_ test tasks is a potential interpretation footgun.

A better idea here would be to configure testing using Test Suites.  But those are still unfortunately `@Incubating`.  Maybe there is some other plugin that adds multiple tasks that also has a DSL that could be used here?  Maybe we could create something that adds a task per X and adds a DSL extension that could be used to only configure some of them?  Maybe we could filter the result of `withType` by adding an `if` in configuration lambda to confirm which tasks we want to configure.

The goal is not to recommend switching blindly to `withTYpe` and setting yourself up for problems when more tasks of that type are added.

## Misc

- Can we also mention "Never use the Groovy `test { /* configuration logic */ }` syntax anywhere" to configure by name, as it has even more potential problems with name shadowing?
- Should we add something about "When adding adhoc tasks always configure them immediately upon registering them to avoid the need to look them up later"?

liutikas commented:
> So I've talked this item over with @big-guy, and there are some clarifications we want to make before getting into more detailed wordsmithing.
> 
> ## The Explanation
> > Task names are consider implementation details of a plugin
> 
> This is sometimes true and sometimes not. Many Gradle core plugins expose the task name as a static constant on the task class or on the plugin class - these would be considered public API and would require a full deprecation cycle if we were ever to change them (this is very unlikely). This may be more true in the Android world, but it's not a universal principle by any means.

It is kind of already broken with core Gradle too. Let's take a specific example of JVM test suites plugin. Adding that means `tasks.named("test")` only gives you one of the test tasks, the default one, the new one just added will not be affected in any way. `tasks.withType<Test>` on the other hand is more robust than the name version of configuring.

Additionally, as far as I can tell outside of lifecycle tasks, Gradle docs have few mentions of what it considers "static constants" and it has almost no real constants in the public API with actual values. It has been mostly keeping these names around by convention and desire not to break users due to years of legacy. 

Finally, even if we don't "often" change it, it scales super poorly to use names for tasks, it is less obvious with core Gradle, but with AGP and KGP where flavors/variants are the norm, it is 10x more important.

> 
> Relatedly, there is no iron law that says that task _types_ have to remain the same from version to version of a plugin. These are also likely to be seen by plugin authors as implementation details in many cases, and might be refactored on a whim. Changes here at least have the advantage of failing fast when a buildscript uses the old name and fails to compile - but this is still annoying.

I am 100% on the same page that even task types are implementation details, that is why I phrased the suggestion to prefer DSL over task types. However, from experience it is *way* better to work with types than names due to increased robustness for variants/flavors and test suites described above.
> 
> > Making assumptions about task names makes a build fragile to Gradle and plugin updates.
> 
> This only applies if names are _not_ part of the plugin's public API.

Where in Gradle docs do you have task names defined as the public API?

The only things i've seen are [`compileJava`](https://docs.gradle.org/current/userguide/java_library_plugin.html)  and [`test`])https://docs.gradle.org/current/userguide/java_testing.html#sec:java_testing_basics) in `java-library` plugin.

If you do have strong convictions you want task names to be considered public API, I think we need a much clearer guidance on which names are considered public, as right now it is all murky.

> 
> There seem to be 2 different personas we're talking to here:
> 
> ### As a Plugin Author:
> Don't force your users to rely on task names _or_ task types. Provide a DSL extension block for configuring behavior that is then wired into the appropriate tasks your plugin creates. This allows consumers of your plugin to work at a higher level of abstraction, using a more declarative approach. Don't expose task names as part of your public API as static constants on task or plugin types.

I'm 100% on the same page on this recommendation. I would say we should go even stronger to have a way for Gradle to mark tasks as invisible/non-queriable outside of the plugin that created them. 
> 
> ### As a Build Author:
> Consider the best way to access a task. Prefer Plugin DSL > Configuring Tasks by Type > Task Names. Be aware that only the first should really be considered future-proof (and when we start writing more plugin author best practices, "Don't Change Your Extensions (including their names) without a deprecation cycle" should be one of them). Sometimes plugin DSL won't be available and you'll have to use types. Sometimes there will be multiple instances of the same type and you'll only want to configure some of them so you might be forced to use names.

Sure, I am mostly on the same page on this as well.
> 
> In an ideal world, a recommendation when running builds might be to only execute _lifecycle tasks_ which lack any configuration except task dependencies. The name and type details of the configurable tasks that actually do work would then be hidden. This would however require Plugin Authors to reliably create such tasks, and wire their own tasks into existing lifecycle tasks as dependencies.

I feel like this is a whole other separate best practice, so please split that out into its own practice. I will not include that in this one.

> 
> That aside, we need to make this item a little more nuanced. We don't want a universal recommendation against calling `named()`.
> 
> ## The Example
> A problem with the example is that the Avoid and Do work completely differently for the `test` configuration example, as noted in the bullet point. In the Avoid, we are intending to configure only a single task. In the Do, we are configuring every task of that type _including any that may be added by other plugins applied later_. These are very different behaviors and assuming this will only configure the _current_ or _desired_ test tasks is a potential interpretation footgun.

"In the Avoid, we are intending to configure only a single task" is an assumption. I think the important thing that ethis example conveys is that most of the time you want *all* tasks of the same type to be configured. If you use junit4, you likely use it everywhere, if you target Java 17, you likely want all tasks to do the same. The point of the example is to point how the flaw in picking a named one as that puts you into false confidence that you have it covered.

> 
> A better idea here would be to configure testing using Test Suites. But those are still unfortunately `@Incubating`. Maybe there is some other plugin that adds multiple tasks that also has a DSL that could be used here? Maybe we could create something that adds a task per X and adds a DSL extension that could be used to only configure some of them? Maybe we could filter the result of `withType` by adding an `if` in configuration lambda to confirm which tasks we want to configure.

I can add that if you find that helpful, but that would add a non-trivial amount of boiler plate to the example
> 
> The goal is not to recommend switching blindly to `withTYpe` and setting yourself up for problems when more tasks of that type are added.

I strongly disagree with this. From experience of maintaining builds I regret inconsistency way more often than accidentally making build too consistent. I think it is way better to think through and catch tasks created by adding new flavors/variants or test suites and still continue to work.
> 
> ## Misc
> * Can we also mention "Never use the Groovy `test { /* configuration logic */ }` syntax anywhere" to configure by name, as it has even more potential problems with name shadowing?

Your docs have this as the top example in the test task, see https://github.com/gradle/gradle/issues/34412

> * Should we add something about "When adding adhoc tasks always configure them immediately upon registering them to avoid the need to look them up later"?

This seems like a completely separate suggestion, please split it out into a separate practice. I will not add it to this practice.

tresat commented:
> Gradle docs have few mentions of what it considers "static constants" and it has almost no real constants in the public API with actual values.

This isn't true - if you open the published public API https://mvnrepository.com/artifact/org.gradle.experimental/gradle-public-api/9.0.0-milestone-9 you can look in org.gradle.api.plugins.JavaPlugin to see a list of task name constants.  We definitely internally consider them to be public API and our infra warns us if we unintentionally change these because we are committed to these names as part of our back compat contracts..  I don't know that we've actually ever changed the name of the test task, for instance.  It's fine to rely on these names for core Gradle plugins - though AGP/KGP might see things differently, for sure.

> Where in Gradle docs do you have task names defined as the public API?  ...  If you do have strong convictions you want task names to be considered public API, I think we need a much clearer guidance on which names are considered public, as right now it is all murky.

The docs could certainly do a better job clarifying what we consider public API, but don't you think that public static final fields on API types are typically recognized as API constants in general?  @Laura Kassovic it might be good to work this clarification into the docs wherever we define public API (and/or in the Best Practice item re: avoiding internals).

> Let's take a specific example of JVM test suites plugin.

I really think we should get away from testing and test suites in this item, since they are an example of a case where we do offer the ability to configure the task without knowing name or type via the DSL: https://docs.gradle.org/current/userguide/jvm_test_suite_plugin.html#sec:configuring_the_built_in_test_suite.  The main problem with demonstrating this abilty is that suites are still incubating :cry:, but this would 100% be the recommendation from us for anyone okay with using an incubating API.

> However, from experience it is way better to work with types than names due to increased robustness for variants/flavors and test suites described above.

I agree in many/most cases it is, especially in the AGP world.  We just think it's important to strongly qualify this as mentioned.

> I would say we should go even stronger to have a way for Gradle to mark tasks as invisible/non-queriable outside of the plugin that created them.

That's sort of the intent for tasks without groups, which don't show up in the tasks report.  It's a very weak mechanism though.

> I feel like this is a whole other separate best practice

Yeah, that was just me musing, not anything actionable for this item.

> I can add that if you find that helpful, but that would add a non-trivial amount of boiler plate to the example

That's really the big problem for demonstrating this item, where the benefits are hard to represent in a simple example build.  I'm honestly not sure what would be best.  Maybe just adding another subproject so that withType could be filtered would be enough?

> I think it is way better to think through and catch tasks created by adding new flavors/variants or test suites and still continue to work. 

Sure, but we have to communicate that, it's why I said "blindly".  As written, this seemed like a universal recommends, which is want we want to be careful to avoid.

> Your docs have this as the top example in the test task

Yeah, we should update these and `init` soon.  I still think its worth mentioning in this item, but perhaps we can just update the item when the other changes are made.

liutikas commented:
I took @lkasso 's suggestion and tweaked it a bit. What are your thoughts on this version @lkasso and @tresat ?

lkasso commented on platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc:
```diff
@@ -409,3 +409,57 @@ The output reveals that the `lib` project is now built when the `:app:goodClassp
+
+  1. Plugin DSL - Future-proof and most stable.
+  2. Task types - Robust in many cases, but can change across plugin versions.
+  3. Task names - Use only when they are explicitly documented as <<public_apis.adoc#public_gradle_apis,public API>>.
```
```suggestion
3. *Task names* - Use only when they are explicitly documented as <<public_apis.adoc#public_gradle_apis,public API>>.
```

lkasso commented on platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc:
```diff
@@ -409,3 +409,57 @@ The output reveals that the `lib` project is now built when the `:app:goodClassp
+<1> *Using java DSL*: Entirely avoids implementation details of `java-library` plugin and sets `sourceCompatibility` for all `JavaCompile` tasks.
+<2> *Using publishing DSL*: Avoids hardcoding task names and ensures that generated pom has the expected URL.
+
+=== Tags
```
```suggestion
=== Tags

```

lkasso reviewed (approved):
Nits but it LGTM overall. It would be nice to get a second opinion.

liutikas committed:
Apply suggestions from code review

Co-authored-by: Laura Kassovic <lkassovic@gradle.com>
Signed-off-by: Aurimas <aurimas@liutikas.net>

tresat commented on platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc:
```diff
@@ -409,3 +409,60 @@ The output reveals that the `lib` project is now built when the `:app:goodClassp
+
+Prefer, in order:
+
+1. *Plugin DSL* - Future-proof and most stable.
```
🤔 It isn't really explained what this means yet.

```suggestion
1. *Plugin DSL* - Configure tasks via extension blocks provided by a plugin that are automatically wired to the tasks that plugin creates.  This is the most future-proof and stable option.
```

tresat commented on platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc:
```diff
@@ -409,3 +409,60 @@ The output reveals that the `lib` project is now built when the `:app:goodClassp
+include::sample[dir="snippets/bestPractices/avoidHardcodingTaskNames-avoid/groovy",files="build.gradle[tags=avoid-this]"]
+====
+
+<1> Using `tasks.named` to get `GenerateMavenPom`: This relies on this task name and that's the only `GenerateMavenPom` task that needs configuring.
```
🤔 This actually relies on name _and type_ explicitly in the Kotlin example (and implicitly in the Groovy one).

How about configuring something common to `Task` so that the type isn't necessary?

```
tasks.named("generatePomFileForMavenPublication").configure { 
    finalizedBy(someOtherTask)
}
```

tresat commented on platforms/documentation/docs/src/docs/userguide/reference/best-practices/best_practices_tasks.adoc:
```diff
@@ -409,3 +409,60 @@ The output reveals that the `lib` project is now built when the `:app:goodClassp
+include::sample[dir="snippets/bestPractices/avoidHardcodingTaskNames-do/groovy",files="build.gradle[tags=ok]"]
+====
+
+<1> Using `tasks.named` to get `JavaCompile`: This uses a public constant to get `JavaCompile`.
```
🤔 I think we should show configuring `javaCompile` in the "avoid" example as well, so that this isn't a huge difference between the avoid and do examples.

Perhaps "avoid" could show using "javaCompile" instead of the constant.  Then in a single "do" snippet, we could show both using the constant and the compatibility DSL.  Looking up this task wouldn't be a case of avoid/do, but rather of ok (inline String)/better (using the constant)/best (DSL).

Even if we don't do that, I think it's kind of jarring to go from reading the avoid example about maven publication to the first recommendation being about java compile when java compile wasn't even present in the avoid.

tresat commented on platforms/documentation/docs/src/snippets/bestPractices/avoidHardcodingTaskNames-avoid/groovy/build.gradle:
```diff
@@ -0,0 +1,20 @@
+        }
+    }
+}
+tasks.named("generatePomFileForMavenPublication") { // <1>
```
```suggestion
}

tasks.named("generatePomFileForMavenPublication") { // <1>
```

tresat reviewed (changes requested):
I'm mostly ready to accept this, and consider my previous reservations as all answered.

I could go either way on the "finalizedBy" suggestion, as it does make the example somewhat less concrete.  I think the others are worth doing, and are relatively minor.

tresat committed:
Update 'Don't hardcode task names' guidance

Expand explanations for configuring tasks via Plugin DSL and task
types. Add a new example for `compileJava` to demonstrate safe
referencing of public API task names using constants. This
enhances the clarity and comprehensiveness of the best practice
guidance.

tresat committed:
Support CC parallelization of independent tasks

tresat committed:
Add project download link to best practices guide

This provides a direct download link for the 'avoidHardcodingTaskNames-do'
example project. This enables users to easily obtain and experiment
with the code snippet discussed in the documentation.
